code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Offer.kt

```diff
@@ -0,0 +1,279 @@
+         */
+        @JvmStatic
+        fun Offer.toEventData(): Map<String, Any> {
+            val offerMap = mutableMapOf<String, Any>()
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_ID] = id
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_ETAG] = etag
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_SCORE] = score
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_SCHEMA] = schema
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_META] = meta
+
+            val data = mutableMapOf<String, Any>()
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_ID] = id
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_TYPE] = type.toString()
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_CONTENT] = content
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_LANGUAGE] = language
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_CHARACTERISTICS] = characteristics
+
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA] = data
+
+            return offerMap
+        }
+
+        private fun getContentFromOfferData(offerData: Map<String, Any?>): String {
+            val offerContent = offerData[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_CONTENT]
+            return when (offerContent) {
+                is String -> offerContent
+                is Map<*, *> -> JSONObject(offerContent).toString()
+                else -> ""
+            }
+        }
+
+        fun List<Offer>?.displayed() {
+            OptimizeUtils.trackWithData(this.generateDisplayInteractionXdm())
+        }
+
+        fun List<Offer>?.generateDisplayInteractionXdm(): Map<String, Any>? {
+            return OptimizeUtils.generateInteractionXdm(
+                OptimizeConstants.JsonValues.EE_EVENT_TYPE_PROPOSITION_DISPLAY,
+                this?.toPropositionsList()
+            )
+        }
+
+        private fun List<Offer>?.toPropositionsList(): List<OptimizeProposition>? {
+            val propositions = this
+                ?.mapNotNull { it.propositionReference?.get() }
+                .takeIf { !it.isNullOrEmpty() }
+            return propositions?.filter { proposition ->
+                proposition.offers.any { offerInProp ->
+                    this?.any { it.id == offerInProp.id } == true
+                }
+            }
+        }
+    }
+
+    var id: String = ""
+        private set
+    var etag: String = ""
+        private set
+    var score: Double = 0.0
+        private set
+    var schema: String = ""
+        private set
+    var meta: Map<String, Any> = HashMap()
+        private set
+    var type: OfferType = OfferType.UNKNOWN
+        private set
+    var language: List<String> = ArrayList()
+        private set
+    var content: String = ""
+        private set
+    var characteristics: Map<String, String> = HashMap()
+        private set
+    var propositionReference: SoftReference<OptimizeProposition>? = null
```

Comment:
There was a public API `getProposition` . With the way its refactored here, that method no longer exists, making this a breaking change

Reply:
You will need to do something like this to achieve parity

```
    @get:JvmName("getPropositionReference")
    @set:JvmName("setPropositionReference")
    internal var propositionReference: SoftReference<OptimizeProposition>? = null
    var proposition = propositionReference?.get()
        private set
```